controllers/auth.go

```diff
@@ -348,33 +377,112 @@ func (ac *AuthController) LoginFinalize(w http.ResponseWriter, r *http.Request)
 		return
 	}
 
-	accountID, err := ac.opaqueService.LoginFinalize(akeStateID, opaqueReq)
+	loginState, err := ac.opaqueService.LoginFinalize(loginStateID, opaqueReq)
 	if err != nil {
 		if errors.Is(err, util.ErrIncorrectCredentials) ||
 			errors.Is(err, util.ErrIncorrectEmail) ||
 			errors.Is(err, util.ErrIncorrectPassword) ||
-			errors.Is(err, util.ErrAKEStateNotFound) ||
-			errors.Is(err, util.ErrAKEStateExpired) {
+			errors.Is(err, util.ErrInterimPasswordStateNotFound) ||
+			errors.Is(err, util.ErrInterimPasswordStateExpired) {
 			util.RenderErrorResponse(w, r, http.StatusUnauthorized, err)
 			return
 		}
+		if errors.Is(err, util.ErrInterimPasswordStateMismatch) {
+			util.RenderErrorResponse(w, r, http.StatusBadRequest, err)
+			return
+		}
 		util.RenderErrorResponse(w, r, http.StatusInternalServerError, err)
 		return
 	}
 
-	session, err := ac.ds.CreateSession(*accountID, datastore.PasswordAuthSessionVersion, r.UserAgent())
+	response := LoginFinalizeResponse{
+		RequiresTwoFA: loginState.RequiresTwoFA,
+	}
+	if !loginState.RequiresTwoFA {
+		// Create a session and return an auth token
+		authToken, err := ac.createSessionAndToken(*loginState.AccountID, r.UserAgent())
+		if err != nil {
+			util.RenderErrorResponse(w, r, http.StatusInternalServerError, err)
+			return
+		}
+		response.AuthToken = &authToken
+	}
+
+	render.Status(r, http.StatusOK)
+	render.JSON(w, r, response)
+}
+
+// @Summary Finalize login with 2FA
+// @Description Final step of 2FA login flow, verifies TOTP code or recovery key and creates a session. If a recovery key is used, 2FA will be disabled.
+// @Tags Auth
+// @Accept json
+// @Produce json
+// @Param Authorization header string true "Bearer + login state token"
+// @Param Brave-Key header string false "Brave services key (if one is configured)"
+// @Param request body services.TwoFAAuthRequest true "2FA verification request"
+// @Success 200 {object} LoginFinalize2FAResponse
+// @Failure 400 {object} util.ErrorResponse
+// @Failure 401 {object} util.ErrorResponse
```

Comment:
Let's note in the swagger docs that bad codes, already-used codes and bad recovery keys will return this status code.